### DISCUSSION TITLE
flow deck v2 patch pcb - Loco, The CF, flow, Big quad

### DISCUSSION THREAD

**Discussion by soulriders** [bitcraze/discussions#263] [Q&A] [answered]
![S20220908_0001](https://user-images.githubusercontent.com/74790243/189040809-bf8405d5-3d36-4528-99b0-78b5035cf63f.jpg)

Hi, 
Anyone knows What point to cut off on the PCB trace?
I cannot findout the gerber file or something for doing this.
Please Help me. 

--------------------------------------------------
To make it work with the Flow (v2) deck more changes/patching are needed:

The Flow HW ChipSelect (IO3) interfere with BigQuad PWM and needs to be moved. Cut trace and solder patch on Flow deck from IO3 to IO1.

**Comment by tobbeanton** [comment, marked as answer]
This is my suggested way of modding it which I think is easiest to restore. The pin end doesn't have to be cut either and can be "pushed" up instead.
![flow-mod](https://user-images.githubusercontent.com/6593186/189323338-faa8edaf-9a4a-4ab4-8b58-6486ee5d2630.png)

If you want to move the Chip Select to another pin change the destination of the jumper wire but still start at IO3.

**Reply by soulriders** [reply]
1LOCO  2CF 3BQD 4FLOW_V2_C1

Conditions.
1. Loco deck uses IO1 for chip select. 
2. FLOW IO3 must be removed from board stacking(circuit) that's because the IO3 noise affects BQD
3. FLOW IO4 must have definition to get Chip secelt on the firmware.  Insted of FLOW IO3 on the code.(flowdeck_v1v2.c)

Process.
1. CF IO3=> wired to CF IO4(vacancy use)
2. FLOW IO4 => fix firmware to make definition (change IO3 into Io4. On the flowdeck_v1v2.c )
3. FLOWIO3 must not be connected to the stack of decks.

**Comment by soulriders** [comment]
Hi,
Thank you for your answer.
I'll try this.
Have a good day.
Best regards,
Gregory.





2022년 9월 9일 (금) 오후 6:55, Tobias Antonsson ***@***.***>님이 작성:

**Comment by soulriders** [comment]
![Screenshot_20220913-150641_Gallery](https://user-images.githubusercontent.com/74790243/189824532-aeb05b83-6c57-4444-9f6d-ee24c451eb79.jpg)

BQD

IOS3 lift up to cut FLOW IO3.

solder a register's leg IO3 IO4 to move signal pin.

**Reply by soulriders** [reply]
fix firmware to make definition (change IO3 into Io4. On the flowdeck_v1v2.c )

**Reply by tobbeanton** [reply]
The way you soldered the pin-header will not work as you will short IO3 and IO4 from the Crazyflie side. You need to do this solder bridge on the Flow deck where it only affects this deck. This is because the IO3 never reaches the flow deck since you pushed this up.

**Comment by soulriders** [comment]
The Power problem is remain.

**Reply by tobbeanton** [reply]
For powering the BQ-deck and CF connect 5V to any connection marked 5V and the same with GND. Using the BUZ 5V pin works but the GND should be connected to something marked GND.

**Reply by soulriders** [reply]
Thank you. I'm trying this.

**Comment by soulriders** [comment]
![image](https://user-images.githubusercontent.com/74790243/190045282-99840459-d929-4886-9a92-f9545140514c.png)
![image](https://user-images.githubusercontent.com/74790243/190045404-7f91b719-55fe-454d-acb3-ee93b5d7177a.png)

**Reply by soulriders** [reply]
wrong.

**Comment by soulriders** [comment]
![image](https://user-images.githubusercontent.com/74790243/190055986-4dc1eeb9-5ead-49f3-a088-c18c355d9cbf.png)

**Comment by soulriders** [comment]
![image](https://user-images.githubusercontent.com/74790243/190058151-fba37e44-fd16-4ab5-bc98-eafffda52b52.png)

**Reply by soulriders** [reply]
FLOW IO4 => fix firmware to make definition (change IO3 into Io4. On the flowdeck_v1v2.c )

**Reply by tobbeanton** [reply]
This is the correct way

**Reply by soulriders** [reply]
Thank you. But not working normally. 
I thinks my case is power problem.

**Comment by soulriders** [comment]
ESTKALMAN: State out of bounds, resetting
HEALTH: Idle:4.91V sag: 0.0V (< 0.95V) [OK]
ESTKALMAN: WARNING: Kalman prediction rate low (95)
STAB: WARNING: stabilizer loop rate is off (947)
ESTKALMAN: State out of bounds, resetting
HEALTH: Acc noise floor variance X+Y:0.028904, (Z:0.980155)
HEALTH: Motor M1 variance X+Y: 0.01 (Z:0.0), voltage sag:0.0
HEALTH: Motor M2 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M3 variance X+Y: 0.01 (Z:0.0), voltage sag:0.0
HEALTH: Motor M4 variance X+Y: 0.01 (Z:0.01), voltage sag:0.0
ESTKALMAN: WARNING: Kalman prediction rate low (82)

**Reply by soulriders** [reply]
error

**Comment by soulriders** [comment]
![image](https://user-images.githubusercontent.com/74790243/190070087-95d8f093-02f8-4d41-bcb7-c1f5dea85b79.png)
not working

**Reply by soulriders** [reply]
Loco, The CF, flow, Big quad

**Reply by tobbeanton** [reply]
The CF2.1 will be powered though the BQ-deck. Don't power it using the battery connector other then with the battery.

**Reply by soulriders** [reply]
I'm sorry again. I've forgot this posting.
I swear I'll keep this community clean.

**Comment by soulriders** [comment]
![image](https://user-images.githubusercontent.com/74790243/190070141-8052c111-c60c-4e96-a146-b05f9caa7f9b.png)
not working

**Reply by soulriders** [reply]
Loco, The CF, flow, Big quad

**Reply by tobbeanton** [reply]
Test with one deck at a time. I suggest starting to get CF+Flow+BQ. What does the console output say in the cfclient after connecting to the CF?

**Reply by soulriders** [reply]
Thank you for your answer. I'll check the output, and will post the message here.

**Reply by soulriders** [reply]
SYS: Build 2:732713b84a7e (2022.09 +2) MODIFIED
SYS: I am 0x20393430533350070046004D and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 3 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x30:0x55
DECK_CORE: Calling INIT on driver bcDWM1000 for deck 2
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
PLATFORM: Low interference mode. NRF51 TX power offset by -12db.
SYS: About to run tests in system.c.
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
Error while initializing the PMW3901 sensor
DECK_CORE: Deck 1 test [FAIL].
DECK_CORE: Deck 2 test [OK].
SYS: deck [FAIL]

**Reply by soulriders** [reply]
CFGBLK: v1, verification [OK]
DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 3 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
DECK_CORE: Calling INIT on driver bcDWM1000 for deck 2
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
PLATFORM: Low interference mode. NRF51 TX power offset by -12db.
SYS: About to run tests in system.c.
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 1 test [OK].
DECK_CORE: Deck 2 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 14936 bytes
STAB: Ready to fly.
ESTKALMAN: State out of bounds, resetting
ESTKALMAN: State out of bounds, resetting

**Reply by soulriders** [reply]
1.BQD motor pwm wired to BQD deck, no power line to CF or BQD
dwm1000 is not working

2. BQD motor PWM wired to bqd deck(I'll call this Motor wired), CF default battery to BQD(I'll call this 2battery)
dwm1000 is working

3. BQD motor not wired, CF default battery wired to BQD
dwm1000 is not working

4. BQD motor wired, no other battery to BQD or CF
CF is working, dwm1000 is not working

**Reply by soulriders** [reply]
I'll make a list for cases below.

**Reply by soulriders** [reply]
![image](https://user-images.githubusercontent.com/74790243/190562374-93452ea5-4f16-4b84-8fab-6fbf19031497.png)

**Reply by soulriders** [reply]
https://docs.google.com/spreadsheets/d/1Z2Awj6MVq4c8fZrfwPG2Yljbb33gSyWwoCD4W_VoGiA/edit#gid=1510422500

**Reply by soulriders** [reply]
Without LOCO case has been started.
It's consist of CF,BQD,FLOW.

**Reply by soulriders** [reply]
CF BQD FLOW case(without LOCO)
1. Battery ON, PDM 5v(4.9v) to BQD
SYS: Crazyflie 2.1 is up and running!
SYS: Build 2:732713b84a7e (2022.09 +2) MODIFIED
SYS: I am 0x20393430533350070046004D and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 1 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 17608 bytes
STAB: Ready to fly.
ESTKALMAN: State out of bounds, resetting
HEALTH: Acc noise floor variance X+Y:0.004561, (Z:0.002784)
HEALTH: Motor M1 variance X+Y: 0.0 (Z:0.0), voltage sag:4.17
HEALTH: Motor M2 variance X+Y: 0.0 (Z:0.0), voltage sag:4.17
HEALTH: Motor M3 variance X+Y: 0.0 (Z:0.0), voltage sag:4.17
HEALTH: Motor M4 variance X+Y: 0.0 (Z:0.0), voltage sag:4.09
STAB: WARNING: stabilizer loop rate is off (77)
ESTKALMAN: WARNING: Kalman prediction rate low (56)
ESTKALMAN: State out of bounds, resetting
HEALTH: Acc noise floor variance X+Y:0.004658, (Z:0.002410)
HEALTH: Motor M1 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M2 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M3 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M4 variance X+Y: 0.01 (Z:0.03), voltage sag:0.0
ESTKALMAN: WARNING: Kalman prediction rate low (77)
ESTKALMAN: State out of bounds, resetting

**Reply by soulriders** [reply]
CF, BQD,FLOW case.
2. only PDM ON (It's powered and connected. I think this is Motor PWM Leakage current .
SYS: Crazyflie 2.1 is up and running!
SYS: Build 2:732713b84a7e (2022.09 +2) MODIFIED
SYS: I am 0x20393430533350070046004D and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 1 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 17608 bytes
STAB: Ready to fly.
ESTKALMAN: State out of bounds, resetting
ESTKALMAN: State out of bounds, resetting
ESTKALMAN: State out of bounds, resetting
HEALTH: Acc noise floor variance X+Y:0.004653, (Z:0.002082)
HEALTH: Motor M1 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M2 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M3 variance X+Y: 0.0 (Z:0.0), voltage sag:0.0
HEALTH: Motor M4 variance X+Y: 0.01 (Z:0.0), voltage sag:0.0
STAB: WARNING: stabilizer loop rate is off (982)
ESTKALMAN: WARNING: Kalman prediction rate low (48)
ESTKALMAN: State out of bounds, resetting
HEALTH: Idle:3.86V sag: 0.0V (< 0.95V) [OK]
ESTKALMAN: WARNING: Kalman prediction rate low (97)
ESTKALMAN: State out of bounds, resetting

**Reply by tobbeanton** [reply]
I will try to replicate your setup and get back to you.

**Reply by tobbeanton** [reply]
In my inital test setup I modified the flow deck the same way you did. I did the following changes to flowdeck_v1v2.c:
```
diff --git a/src/deck/drivers/src/flowdeck_v1v2.c b/src/deck/drivers/src/flowdeck_v1v2.c
index bfd971ec..8431f116 100755
--- a/src/deck/drivers/src/flowdeck_v1v2.c
+++ b/src/deck/drivers/src/flowdeck_v1v2.c
@@ -78,7 +78,7 @@ static bool useAdaptiveStd = false;
 // (will not work if useAdaptiveStd is on)
 static float flowStdFixed = 2.0f;
 
-#define NCS_PIN DECK_GPIO_IO3
+#define NCS_PIN DECK_GPIO_IO4
 
 
 static void flowdeckTask(void *param)
@@ -244,7 +244,7 @@ static const DeckDriver flowdeck2_deck = {
   .pid = 0x0F,
   .name = "bcFlow2",
 
-  .usedGpio = DECK_USING_IO_3,
+  .usedGpio = DECK_USING_IO_4,
   .usedPeriph = DECK_USING_I2C | DECK_USING_SPI,
   .requiredEstimator = kalmanEstimator,
```
I enabled BigQuad capabilities in kbuild. Outdated documentation. See [this question](https://github.com/orgs/bitcraze/discussions/269)
I only power the setup using the battery connected the normal way as a first step. It works with the following console output:
```
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 53:c1ed80cc4a88 (2022.05-rc2 +53) MODIFIED
SYS: I am 0x203937434848500F001E0035 and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 3 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
DECK_CORE: Calling INIT on driver bcBigQuad for deck 1
BIGQUAD: Switching to brushless.
MTR-DRV: Using brushless motor driver
EXTRX: CPPM initialized, expecting AETR channel mapping
DECK_CORE: Calling INIT on driver bcDWM1000 for deck 2
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
PLATFORM: Low interference mode. NRF51 TX power offset by -12db.
SYS: About to run tests in system.c.
SYS: NRF51 version: 2022.05
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
DECK_CORE: Deck 1 test [OK].
DECK_CORE: Deck 2 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 14944 bytes
EXTRX: CPPM Task Started
STAB: Ready to fly.
```

**Reply by tobbeanton** [reply]
This also works well when powering the BQD at a 5V + GND input. The voltage is 4.97V.

**Reply by tobbeanton** [reply]
![20220916_113700](https://user-images.githubusercontent.com/6593186/190608348-16ebb2c1-9376-443f-b34c-c0ce75176d35.jpg)
![20220916_113715](https://user-images.githubusercontent.com/6593186/190608379-3faa6298-e586-4ab8-b13c-1615397706f0.jpg)
